File: leetcode-04/solutions/cpp/0109.cpp
class Solution {
 public:
  TreeNode* sortedListToBST(ListNode* head) {
    int length = 0;
    for (auto curr = head; curr; curr = curr->next) length++;

    this->head = head;

    return helper(0, length - 1);
  }

 private:
  ListNode* head;

  TreeNode* helper(int l, int r) {
    if (l > r) return NULL;

    int mid = (l + r) >> 1;

    TreeNode* left = helper(l, mid - 1);
    TreeNode* node = new TreeNode(head->val);
    head = head->next;
    node->left = left;
    node->right = helper(mid + 1, r);

    return node;
  }
};

File: leetcode-04/solutions/python3/0109.py
class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        def helper(l, r):
            nonlocal head

            if l > r:
                return None

            mid = (l + r) >> 1

            left = helper(l, mid - 1)
            node = TreeNode(head.val)
            head = head.next
            node.left = left
            node.right = helper(mid + 1, r)

            return node

        length = 0
        curr = head
        while curr:
            length += 1
            curr = curr.next

        return helper(0, length - 1)
